Added ferromagnetic models and glauber sampling functions

forrestsheldon opened the pull request.

Commit:
Added ferromagnetic models and glauber sampling functions

ccoffrin reviewed (commented):
LGTM.  We should find out why the tests are not passing and fix that.  My guess is the typo correction for `variable_count`.  You can test the tests locally with `] test GraphicalModelLearning` this run the `test/runtests.jl` script to make sure everything in the package is working.

Commit:
Incorporated comments from previous pull request

ccoffrin commented on src/models.jl:
```diff
@@ -244,3 +270,99 @@ function permutations(partial_perm::Array{Any,1}, items, order::Int, asymmetric:
+
+    for (interacting, weight) in gm.terms
+        for (idx, site) in enumerate(interacting)
+            push!(neighbors[site], ([interacting...], weight))
```
Ok, yes a name change sounds good to me.  With the docs clarified this is fine as it.

forrestsheldon reviewed (commented):
* A missing return line in a convert(...) was causing the tests to fail. 
* The comment has been moved outside of the function and encapsulated in triple quotes.
* For the neighbors function, I made the change but can't think of a useful case in which a graph would contain a node that is not connected to anything.  Do we expect these cases?

Commit:
Changed name of generate_neighbors to generate_neighborhoods

ccoffrin commented:
In terms of correctness and style this is good to go.   The best practice would be to also add a test so that these features don't break in future versions, but this can wait if this code is still in flux.  I'll leave that determination to you.  Merge this into master when you are ready.

ccoffrin reviewed (approved):
LGTM.

Commit:
Added sample method for Glauber sampling and an entropic descent learning method

Commit:
Modified entropic descent to output objectives until we decide on termination criteria

Commit:
Added gradient termination tolerance option and warning in case optimization terminates due to maximum steps

Commit:
Merge remote-tracking branch 'origin/master' into glauber

Commit:
Add return_objectives option to learn with entropic descent for debugging.

Commit:
Added structured learning functions to GML and some helper functions to models

Commit:
Added batching to the Glauber sampling method

Commit:
Added batch sampling methods to reduce memory usage

Commit:
fixed problems with batch and conditional sampling

Commit:
Streamlined glauber sampling routines and added structured learning for models with arbitrary order interactions

Commit:
Added sampling routine that stores spin products in neighborhoods and a learning routine to use them

Commit:
Added dependency on Random to use shuffle function

forrestsheldon committed:
Improving sampling

Added methods to store only local configurations of the state, drastically reducign sample storage.

Commit:
Simplifying neighbor sampling

forrestsheldon committed:
Added sample subset and supporting functions

Commit:
Fixed single site reconstruction

Commit:
Minor corrections to sample_subset and learn_temperature

Commit:
Learn temperature added and sample now can return the final state

Commit:
Revised sampling functions to all return dictionaries.  conversion to array is straightforward and can be done by hand

Commit:
small change to kwargs in samplesubset

forrestsheldon committed:
Simplified sampling to always give dictionaries and cleaned up aspects of model

Commit:
Added several new models

forrestsheldon committed:
Some small notation changes in models.jl

Commit:
add back Project.toml

forrestsheldon committed:
Merge branch 'master' into glauber

forrestsheldon committed:
Added Random to dependencies
